### PR TITLE
[clap-v3-utils] Add support for ElGamal keypair and authenticated encryption key

### PR DESCRIPTION
**Pull request by samkim-crypto**
#### Problem
The clap utils do not yet support ElGamal keypairs and authenticated encryption key that are needed in the token-cli for the token-2022 program.

#### Summary of Changes
- To minimize the repeat of the function `keypair_from_path` for ElGamal and authenticated encryption keys, the `keypair_from_path` function was updated to use generic `EncodableKey` trait types.
-  A new trait `EncodableKey` was defined in the sdk that defines the interface for which cryptographic keys/keypairs can be read, written, and derived from sources.
- `Keypair`, `ElGamalKeypair` and `AeKeys` updated to implement `EncodableKey`.
- Updated/added some docs and errors types in the `zk-token-sdk/encryption` module

Fixes #

### CODE REVIEW

File: sdk/src/signer/mod.rs [hunk 1]
```diff
@@ -103,6 +109,25 @@ pub fn unique_signers(signers: Vec<&dyn Signer>) -> Vec<&dyn Signer> {
     signers.into_iter().unique_by(|s| s.pubkey()).collect()
 }
 
+/// The `EncodableKey` trait defines the interface by which cryptographic
+/// keys/keypairs are read, written, and derived from sources.
+pub trait EncodableKey: Sized {
```
Comment (t-nelson) [inline sdk/src/signer/mod.rs:114]: can you please squash all changes related to the introduction of this trait and its implementation on `Keypair` down to isolated commits and factor them out to a separate "foundational" PR?

File: clap-v3-utils/src/keypair.rs [hunk 2]
```diff
@@ -370,22 +370,22 @@ impl DefaultSigner {
     }
 }
 
-pub(crate) struct SignerSource {
-    pub kind: SignerSourceKind,
+pub(crate) struct Source {
+    pub kind: SourceKind,
```
Comment (t-nelson) [inline clap-v3-utils/src/keypair.rs:374]: are these renames strictly necessary?  the existing types are part of the public API, so even if so, we need to keep the old names around as deprecated symbols.

if not required, please revert.  otherwise, please factor the changes out to a separate PR for discussion

**Review by t-nelson — CHANGES_REQUESTED** [review]
this is pretty big and has some pretty clear boundaries to be split up into smaller PRs. suggestions below

**Comment by CriesofCarrots** [comment]
I was just going to look at this, but will hold off given Trent's comments. Please give me a nudge for split PRs!

**Comment by samkim-crypto** [comment]
Great! I will create separate PRs for these. Thanks for taking a look!

File: clap-v3-utils/src/keypair.rs (hunk 2 above)
Comment (samkim-crypto) [reply clap-v3-utils/src/keypair.rs:374]: Yeah, the renames are not necessary and it is just to have better names since we are now also deriving encryption keys, which are not signers (`SignerSource` --> `Source`). I wouldn't have made the changes if `SignerSource` had public visibility, but it seemed like it was limited to the crate only. But maybe it is better to have consistency between `clap-utils` and `clap-v3-utils`, so I will revert.

**Comment by samkim-crypto** [comment]
Subsumed by https://github.com/solana-labs/solana/pull/30947 and https://github.com/solana-labs/solana/pull/31496.
